File: 5. Longest Palindromic Substring.py
# timeout

class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        res = s[0]
        T = [[False for j in xrange(len(s))] for i in xrange(len(s))]
        for i in xrange(len(s)):
            for j in xrange(i):
                T[j][i] = (s[j] == s[i] and (i-j < 2 or T[j+1][i-1]))
                if T[j][i]:
                    if i-j+1 > len(res):
                        res = s[j:i+1]
            T[i][i] = 1
        return res

File: 5. Longest Palindromic Substring.cpp
// timeout

class Solution {
public:
    string longestPalindrome(string s) {
        string res = "";
        if (s.length() > 0) res = s[0];
        vector <vector <bool>> T (s.length(), vector <bool> (s.length(), false));
        for (int i = 0; i < s.length(); i++)
        {
            for (int j = 0; j < i; j++)
            {
                T[j][i] = (s[j] == s[i] && (i-j<2 || T[j+1][i-1]));
                if (T[j][i] && res.length() < i-j+1)
                    res = s.substr(j, i-j+1);
            }
            T[i][i] = true;
        }
        return res;
    }
};
